pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -0,0 +1,336 @@
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	_, log := logger.GetNewContextWithLogger()
+
+	// Create a new controller
+	c, err := controller.New("csinodetopology-controller", mgr, controller.Options{Reconciler: r,
+		MaxConcurrentReconciles: defaultMaxWorkerThreadsForCSINodeTopology})
+	if err != nil {
+		log.Errorf("Failed to create new CSINodetopology controller with error: %+v", err)
+		return err
+	}
+
+	// Initialize backoff duration map
+	backOffDuration = make(map[string]time.Duration)
+
+	// Predicates are used to determine under which conditions
+	// the reconcile callback will be made for an instance.
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// The CO calls NodeGetInfo API just once during the node registration,
+			// therefore we do not support updates to the spec after the CR has been reconciled.
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Instances are deleted by the garbage collector automatically after the corresponding NodeVM is deleted.
+			// No reconcile operations are required.
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource CSINodeTopology
+	err = c.Watch(&source.Kind{Type: &csinodetopologyv1alpha1.CSINodeTopology{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		log.Errorf("Failed to watch for changes to CSINodeTopology resource with error: %+v", err)
+		return err
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileCSINodeTopology implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCSINodeTopology{}
+
+// ReconcileCSINodeTopology reconciles a CSINodeTopology object
+type ReconcileCSINodeTopology struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client        client.Client
+	scheme        *runtime.Scheme
+	clusterFlavor cnstypes.CnsClusterFlavor
+	configInfo    *cnsconfig.ConfigurationInfo
+	recorder      record.EventRecorder
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Note: The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileCSINodeTopology) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logger.GetLogger(ctx)
+
+	// Fetch the CSINodeTopology instance
+	instance := &csinodetopologyv1alpha1.CSINodeTopology{}
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("CSINodeTopology resource with name %q not found. Ignoring since object must have been deleted.", request.Name)
```

Comment:
In the predicate function on line 137, you're not queueing requests for objects that are deleted.